### DISCUSSION TITLE
Generate Social-Images via API

### DISCUSSION THREAD

**Discussion by Ryuno-Ki** [mdn/yari#3638] [Ideas]
This spun off in https://github.com/mdn/yari/issues/3590#issuecomment-826308746

The dev.to platform is using for https://docs.htmlcsstoimage.com/ - which could be an option for generating images customized to MDN pages.

**Comment by peterbe** [comment]
I don't know who would have an opinion on this. Would be greater-than-zero nice to have but is it important? 
@HerminaC @chrisdavidmills ?

The topic at hand is that we can make the preview prettier on things like Twitter and Facebook if someone posts an MDN URL. 
On my personal blog, I usually take a screenshot of some code (because so many of my personal blog posts are about coding) and use that as an image. Another option would be to generate a screenshot from the top portion of the document at some decent ratio (e.g. 16:9). Or, a third option is to do what dev.to does which is to generate an image on the fly based on some text. 

Another option is to do nothing. Perhaps how the preview of a URL looks like in Twitter isn't greatly important to our traffic.

**Comment by chrisdavidmills** [comment]
I don't think this is hugely important, but it'd be nice to get this sorted with minimal fuss.

Since most of our pages don't have nice screenshots to adapt, can we just do something like have an MDN logo that gets used for all social cards, and just put the document title above it?

**Reply by Ryuno-Ki** [reply]
So I picked the image from https://github.com/mdn/yari/pull/3621#issuecomment-827687130 and the first sentence from https://developer.mozilla.org/en-US/docs/Web/JavaScript and went to https://htmlcsstoimage.com/demo

With a bit of markup

```html
<div class="p-3 text-center mt-3" style="width: 500px">
  <span class="tweet-text mb-3">
    JavaScript (JS) is a lightweight, interpreted, or just-in-time compiled programming language with first-class functions.
  </span>
  <div>
    <img src="https://user-images.githubusercontent.com/413984/116264223-164d0e80-a7b5-11eb-8daf-18f85204ebb0.png" class="rounded-circle shadow border mt-4" width="100px">
  </div>
  <h4 class="mt-2">
    MDN WebDocs
  </h4>
  <span class="text-muted">@MozDevNet</span>
</div>

<!-- Include external CSS, JavaScript or Fonts! -->
<link href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk" crossorigin="anonymous">

<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700" rel="stylesheet">
```

I could generate

![hcti](https://user-images.githubusercontent.com/3097194/116284522-74fe9200-a78d-11eb-88c3-b0ed0920afde.png)

(Ignore the wrong font for a moment).

@chrisdavidmills Be aware of the pricing table at https://htmlcsstoimage.com/pricing

**Reply by peterbe** [reply]
> I don't think this is hugely important, but it'd be nice to get this sorted with minimal fuss.
> 
> Since most of our pages don't have nice screenshots to adapt, can we just do something like have an MDN logo that gets used for all social cards, and just put the document title above it?

Yeah, we can just use the dino logo. That would be minimal fuss. But then it's just a logo which might make the card stand out a whee bit more in the Twitter browsing if you care about that.

**Reply by tannerdolby** [reply]
> Be aware of the pricing table at https://htmlcsstoimage.com/pricing

@Ryuno-Ki I think this is the kicker, the "starter" account that is free only offers up to 50 images and MDN has many more than 50 images needing to be processed so it might be an unnecessary cost to use https://htmlcsstoimage.com, on the other hand it could prove to be the most "bang for your buck" as opposed to an in-house custom implementation.

I stumbled upon [sindresorhus/capture-website](https://github.com/sindresorhus/capture-website) and tested it locally. It works really well to capture a screenshot of a webpage or pass in straight up HTML with internal or external css at build time e.g. `captureWebsite.file(URL or HTML, "<output-file-name>", { config })`, comes with many config options for selecting certain areas of the viewport using selectors, and some other cool stuff.. This could take the same HTML @Ryuno-Ki shows above and pass it into `captureWebsite.file()` to produce a social share image in a similar fashion.

Either method could be useful for generating social images for the docs pages that arent SPAs which would make more sense to display the title, MDN logo, twitter handle etc like shown above rather than only the generic dino logo or MDN svg from mozilla hacks.

**Reply by tannerdolby** [reply]
@peterbe I pulled into your feature branch and added a `<meta name="og:image">` tag along with a `<meta name="twitter:image">` to the HTML and supplied `content=""` with the [MDN Web Docs](https://hacks.mozilla.org/files/2020/10/mdn-web-docs-vector-logo.png) svg logo.

It might end up being more fuss to deal with the title/dino logo/twitter handle for the social image, when we could use this logo (or dino) for pages like mozilla hacks does. With the logo shown below plus the `<title>` and `<meta name="description">` tags the summary card looks really nice.

<img width="1031" alt="Screen Shot 2021-04-29 at 3 30 24 PM" src="https://user-images.githubusercontent.com/48612525/116626717-97fd8300-a900-11eb-8018-1f12d7702b05.png">

<img width="1082" alt="Screen Shot 2021-04-29 at 3 31 16 PM" src="https://user-images.githubusercontent.com/48612525/116626727-9af87380-a900-11eb-80c1-bd34e32f5f6f.png">

**Reply by peterbe** [reply]
@tannerdolby Not sure what your point is here. Does it work, using the `og:*` if you just add the `og:image` **and** `twitter:image`?

**Reply by tannerdolby** [reply]
@peterbe My point was to illustrate how the summary card would look if we **only**  add the missing `<meta name="og:image" content="">` tag to your PR along with `twitter:card`, no need for `twitter:image`.

When the `<meta name="twitter:card" content="">` tag is provided the twitter processor correctly falls back to the `og:*` tags without having to include the other `twitter:*` tags like you originally shared with us from the docs.

**Reply by schalkneethling** [reply]
https://github.blog/2021-06-22-framework-building-open-graph-images/

**Reply by tannerdolby** [reply]
^ Super informative article, thanks for sharing @schalkneethling !
